_posts/2018-07-23-DroolsDataQualityChecking.md

```diff
@@ -0,0 +1,153 @@
+	Collection<Component> objects = (Collection<Component>) kSession.getObjects();
+
+	//marking session for garbage colleciton
+	kSession.dispose();
+
+	//and returning processed objects
+	return objects;
+}
+```
+
+That's all! 
+
+Now we can take a look of heart and soul of drools powered applications - rules. Rules are made using "native" rule language and placed typically into a file with a .drl extension.
+
+A rule has following structure:
+
+```
+rule "name"
+    attributes
+    when
+        condition 
+    then
+        action
+end
+```
+
+Several rules may be placed into one rule file.
+
+
+### 1. Check component type.
+Every component we check should belong to one of two types "GENERIC" or "METER"
+however we may find other types in data we get. Let's check it with one simple drools rule.
+
+this rule will trigger on every object that has something else than "GENERIC" or "METER" in its type property.
+
+```
+rule "Type check"
+    when //condition           
+        $component : Component(type not in ("METER", "GENERIC"))
+    then //action to take
+        $component.messages.add("Type not allowed");
+end
+```
+
+For every object that satisfies condition an action will be taken. That's all, as simple as it is.
+
+
+### 2. Find date and fix date format.
+
+This rule will be applied to every object that has something that sounds like "installed" in it's metadata (Yeah, drools may find things using `soundslike` keyword) and convert value to SQL compliant format. This rule uses external function written in Java. To be able to use java functions inside of drl files we can import them with "import" keyword, as in Java.
+```
+rule "Installed date check"
+    when
+        component : Component($metadata: metadata)
+        entry : Entry() from $metadata.entrySet()
+        key : String() from entry.getKey()
+        value: String() from entry.getValue()
+        Boolean(booleanValue == true) from key soundslike  "installed"
+    then
+        $metadata.put(key, DateUtils.convertDateFormat(value));
+end
+```
+
+### 3. Validate object id
+
+In our application every component has an id expressed as string that starts with 111 if component has type "METER" or 222 if object has type "GENERIC" and we want find all components that violate this naming schema.
+
+```
+rule "ID-Type validation"
+    when 
+        $component : Component(
+            (type == "METER" && id not matches "111*") ||
+            (type == "GENERIC" && id not matches "222*")
+        )
+    then
+        $component.messages.add("ID invalid");
+end
+
+```
+
+# Conclusion
+
+In this article we covered only one little bit of functionality provided by Drools rule engine that is still enough to make rules for data validation and conversion. More complicated rules can do everything that usual Java code can do but with full separation from application logic. So, you don't need redeploy your application every time when rules change.
```

Comment:
What does the last sentence mean? You won't need to recompile only if your rules are an external ref right? That's not explicit.

Reply:
Also, I don't see where is the reference to the rule file.

Reply:
added code for rule loading, explained hot reload more explicit